Initial implementation of parseCrashes()

arturocuya opened the pull request:
The base tool structures only the paths found in the crash report. The goal of this PR is to also structure the remaining information for each crash:

* The error message
* The hardware platforms where the crash occured and their count. ~Not only the code name but also additional information (taken from [here](https://developer.roku.com/es-pe/docs/specs/hardware.md)) like the product name and the model, e.g: Roku Ultra, model 4800X.~
* The application versions and the error count for each.
* The stacktrace, with the scope names and locations of each scope level
* The local variable names and their metadata

This structured list of crashes is stored in `CrashlogFile.crashes`. Now `CrashlogFile.crashes` and `CrashlogFile.references` have redundant information, but I have not removed the latter because I don't know much about the codebase outside of `CrashlogFile`.

I have left some TODOs with questions. Please address them adding comments to the PR.

TwitchBronBron commented on src/CrashlogFile.spec.ts:
```diff
@@ -4,6 +4,9 @@ import { util as bscUtil, standardizePath as s } from 'brighterscript';
 import { CrashlogFile } from './CrashlogFile';
 import { Runner } from './Runner';
 import type { RunnerOptions } from './interfaces';
+// TODO: How to remove this? No quickfix.
```
You just have to move the line up and down until it's sorted alphabetically and the error goes away.

TwitchBronBron commented on src/CrashlogFile.spec.ts:
```diff
@@ -137,6 +140,255 @@ describe('CrashlogFile', () => {
+
+            it('contains underscores', () => {
+                expectContainSubset(
+                    // eslint-disable-next-line no-tabs
```
You can use `\t` instead a literal tap to acomplish this (and then remove the eslint disable comment)

TwitchBronBron commented on src/CrashlogFile.ts:
```diff
@@ -1,8 +1,25 @@
+    StackTrace = 'StackTrace'
+}
+
+const HardwarePlatformHeaderRegex = /\s*count\s+Hardware Platform/;
```
The javascript engine is very good at caching regex literals, so you don't need to lift these into their own variables. Can you move these inline with where they're used?

TwitchBronBron commented on src/CrashlogFile.ts:
```diff
@@ -79,6 +104,107 @@ export class CrashlogFile {
+
+        // TODO: redundant code from StandardReporter, how to refactor?
+        // walk the references backwards and replace the file contents
+        for (const ref of [...this.references].reverse()) {
```
Why is this loop here?

arturocuya replied:
@TwitchBronBron I saw that `CrashlogFile.contents` does not include the mapped source locations generated after `parse()`. This replacement is done in `StandardReported.getFileContents()`. But I want the replacement done before I generate the crashes structure. So the for loop is adapted from `StandardReported.getFileContents()`. But now the code is redundant. What should I do?

TwitchBronBron commented on src/CrashlogFile.ts:
```diff
@@ -1,8 +1,25 @@
+import { util } from './util';
+
+// TODO: where to put this
+enum CrashReportSectionType {
```
I'm fine with it being declared in this file. Can you move it to the bottom though?

TwitchBronBron commented on src/util.ts:
```diff
@@ -13,6 +14,289 @@ class Util {
+    /**
+     * @see https://developer.roku.com/docs/specs/hardware.md
+    */
+    public getAllRokuPlatforms(): RokuHardwarePlatform[] {
```
@chrisdp do you think it would be worth publishing a separate node_module with just this type of information? We could try to keep it up-to-date as new devices come out, and that way all of this type of information would be centralized in a single location.

TwitchBronBron commented on src/util.ts:
```diff
@@ -13,6 +14,289 @@ class Util {
+                productName: 'Roku Express',
+                codeName: 'Nemo',
+                model: '3930X',
+                productionStatus: ProductionStatus.Current
```
Is there a benefit to us storing the end-of-life date on this rather than a hardcoded production status? Just thinking out loud, these statuses will need to be manually changed in this package as the years progress. Perhaps that wouldn't get us any benefit?

TwitchBronBron replied:
Nevermind

arturocuya commented:
@TwitchBronBron What about the redundant information in `CrashlogFile.crashes` and `CrashlogFile.references`?

TwitchBronBron commented on src/CrashlogFile.ts:
```diff
@@ -79,6 +89,106 @@ export class CrashlogFile {
+                } else {
+                    if (foundSectionHeader && currentSection && !/\s*---+\s*/.exec(line)) {
+                        // TODO: What's this?
+                        // eslint-disable-next-line @typescript-eslint/no-loop-func
```
This is an annoying rule. Go ahead and disable it in `eslintrc.json` and then remove these comments.

arturocuya replied:
Done on the latest commit.

TwitchBronBron commented:
> @TwitchBronBron What about the redundant information in `CrashlogFile.crashes` and `CrashlogFile.references`?

There might be situations where there are pkg paths not found inside a crash. The `references` collection handles every pkgPath in the entire crashlog. 

Perhaps for `CrashlogFile.crashes`, you could add a property called `location` that points to the specific `CrashlogFile.references` entry. So instead of `crash.pkgPath`, you would do `crash.location.pkgPath`. That way, you get the sourcemap lookups for free since that `location` would be a reference to the existing `CrashlogFile.references` item.

arturocuya commented:
@TwitchBronBron Update:

- Improved stack trace parsing and added tests for it.
- Removed unnecesary loop.
- Turned off linter rule @typescript-eslint/no-loop-func
- Renamed StackTraceStep to StackFrame
- Replaced `StackFrame`'s `pkgLocation` and `srcLocation` for `reference` which is a pointer to a `CrashlogFile.references` entry.

arturocuya commented on src/CrashlogFile.ts:
```diff
@@ -125,4 +219,192 @@ export class CrashlogFile {
+
+                const match = /(\w+:\/.*?)\((\d+)\)/.exec(pkgLocationAsArray.join(' ').trim());
+                if (match) {
+                    // We need the range to calculate the reference index. To generate the range, we need the line number.
```
@TwitchBronBron I honestly couldn't think of a more elegant solution for this. I wanted to make the subparsers for each crash section agnostic to the order of the sections, so Roku can miss sending, for example, the hardware information and the parser could still figure out the stack trace. This resulted in the subparser for the stack trace not having the original line number for each of its section lines. Hence this code...

arturocuya commented on src/interfaces.ts:
```diff
@@ -57,3 +82,43 @@ export interface Location {
     generate(runner: Runner): any;
 }
+
+export interface StackTraceStep {
```
Done on the latest commit.

TwitchBronBron reviewed (approved):
Everything looks good here. Is there anything else to do before I merge this?

arturocuya commented:
> Everything looks good here. Is there anything else to do before I merge this? The only thing I can see is to get the CI failures fixed.

@TwitchBronBron Yep, just wait for me to fix the coverage problem.
